README.rst
Configuration File Support for Click Commands
===============================================================================

.. image:: https://github.com/click-contrib/click-configfile/actions/workflows/test.yml/badge.svg
    :target: https://github.com/click-contrib/click-configfile/actions/workflows/test.yml
    :alt: CI Build Status

.. image:: https://img.shields.io/pypi/v/click-configfile.svg
    :target: https://pypi.org/project/click-configfile
    :alt: Latest Version

.. image:: https://img.shields.io/pypi/dm/click-configfile.svg
    :target: https://pypi.org/project/click-configfile
    :alt: Downloads

.. image:: https://img.shields.io/pypi/l/click-configfile.svg
    :target: https://pypi.org/project/click-configfile/
    :alt: License


click_ is a Python package for creating beautiful command line interfaces
in a composable way with as little code as necessary.

This package extends the click_ functionality by adding support for commands
that use configuration files.

.. _click: https://click.pocoo.org/


EXAMPLE:

.. code-block:: python

    # -- FILE: example_command_with_configfile.py (PART 1)
    # BASIC SOLUTION FOR: Command that uses one or more configuration files.
    import click

    CONTEXT_SETTINGS = dict(default_map=ConfigFileProcessor.read_config())

    @click.command(context_settings=CONTEXT_SETTINGS)
    @click.option("-n", "--number", "numbers", type=int, multiple=True)
    @click.pass_context
    def command_with_config(ctx, numbers):
        """Example for a command that uses an configuration file"""
        pass
        ...

    if __name__ == "__main__":
        command_with_config()

The command implementation reads the configuraion file(s) by using the
``ConfigFileProcessor.read_config()`` method and stores it in the
``default_map`` of the ``context_settings``.

That is only the first part of the problem. We have now a solution that allows
us to read configuration files (and override the command options defaults)
before the command-line parsing begins.

In addition, there should be a simple way to specify the configuration schema
in the configuration file in a similar way like the command-line options.
An example how this functionality may look like, is shown in the following
code snippet:

.. code-block:: python

    # -- FILE: example_command_with_configfile.py (PART 2)
    # Description of sections in a confguration file: *.ini
    from click_configfile import matches_section, Param, SectionSchema

    class ConfigSectionSchema(object):
        """Describes all config sections of this configuration file."""

        @matches_section("foo")
        class Foo(SectionSchema):
            name    = Param(type=str)
            flag    = Param(type=bool, default=True)
            numbers = Param(type=int, multiple=True)
            filenames = Param(type=click.Path(), multiple=True)

        @matches_section("person.*")   # Matches multiple sections
        class Person(SectionSchema):
            name      = Param(type=str)
            birthyear = Param(type=click.IntRange(1990, 2100))



The example shows that the ``Param`` class supports similar arguments like a
``click.Option``. You can specify:

* a ``type`` (converter), like in click_ options
* a ``multiple`` flag that is used for sequences of a type
* an optional ``default`` value (if needed or used as type hint)

An example for a valid configuration file with this schema is:

.. code-block:: INI

    # -- FILE: foo.ini
    [foo]
    flag = yes      # -- SUPPORTS: true, false, yes, no (case-insensitive)
    name = Alice and Bob
    numbers = 1 4 9 16 25
    filenames = foo/xxx.txt
        bar/baz/zzz.txt

    [person.alice]
    name = Alice
    birthyear = 1995

    [person.bob]
    name = Bob
    birthyear = 2001


The following code snippet shows the remaing core implementation of reading
the configuration file (and parsing the configuration file data):

.. code-block:: python

    # -- FILE: example_command_with_configfile.py (PART 3)
    import configparser     # HINT: Use backport for Python2
    from click_configparser import generate_configfile_names, \
        select_config_sections, parse_config_section

    class ConfigFileProcessor(object):
        config_files = ["foo.ini", "foo.cfg"]   # Config filename variants.
        config_sections = ["foo", "person.*"]   # Sections of interest.
        config_section_schemas = [
            ConfigSectionSchema.Foo,
            ConfigSectionSchema.Person,
        ]

        # -- GENERIC PART:
        # Uses declarative specification from above (config_files, config_sections, ...)
        @classmethod
        def read_config(cls):
            configfile_names = list(generate_configfile_names(cls.config_files))
            print("READ-CONFIG: %s" % repr(configfile_names))
            parser = configparser.ConfigParser()
            parser.optionxform = str
            parser.read(configfile_names)

            storage = {}
            for section_name in select_config_sections(parser.sections(),
                                                       cls.config_sections):
                config_section = parser[section_name]
                cls.process_config_section(config_section, storage)
            return storage

        # -- SPECIFIC PART:
        # Specifies which schema to use and where data should be stored.
        @classmethod
        def process_config_section(cls, config_section, storage):
            """Process the config section and store the extracted data in
            the param:`storage` (as outgoing param).
            """
            if not storage:
                # -- INIT DATA: With default parts.
                storage.update(dict(_PERSONS={}))

            if config_section.name == "foo":
                schema = ConfigSectionSchema.Foo
                section_data = parse_config_section(config_section, schema)
                storage.update(section_data)
            elif section_name.startswith("persons."):
                person_name = section_name.replace("person.", "", 1)
                schema = ConfigSectionSchema.Person
                section_data = parse_config_section(config_section, schema)
                storage["_PERSONS"][person_name] = section_data
            # -- HINT: Ignore unknown section for extensibility reasons.


The source code snippet above already contains a large number of generic
functionality. Most of it can be avoided for processing a specific
configuration file by using the ``ConfigFileReader`` class.
The resulting source code is:

.. code-block:: python

    # MARKER-EXAMPLE:
    # -- FILE: example_command_with_configfile.py (ALL PARTS: simplified)
    from click_configfile import ConfigFileReader, Param, SectionSchema
    from click_configfile import matches_section
    import click

    class ConfigSectionSchema(object):
        """Describes all config sections of this configuration file."""

        @matches_section("foo")
        class Foo(SectionSchema):
            name    = Param(type=str)
            flag    = Param(type=bool, default=True)
            numbers = Param(type=int, multiple=True)
            filenames = Param(type=click.Path(), multiple=True)

        @matches_section("person.*")   # Matches multiple sections
        class Person(SectionSchema):
            name      = Param(type=str)
            birthyear = Param(type=click.IntRange(1990, 2100))


    class ConfigFileProcessor(ConfigFileReader):
        config_files = ["foo.ini", "foo.cfg"]
        config_section_schemas = [
            ConfigSectionSchema.Foo,     # PRIMARY SCHEMA
            ConfigSectionSchema.Person,
        ]

        # -- SIMPLIFIED STORAGE-SCHEMA:
        #   section:person.*        -> storage:person.*
        #   section:person.alice    -> storage:person.alice
        #   section:person.bob      -> storage:person.bob

        # -- ALTERNATIVES: Override ConfigFileReader methods:
        #  * process_config_section(config_section, storage)
        #  * get_storage_name_for(section_name)
        #  * get_storage_for(section_name, storage)


    # -- COMMAND:
    CONTEXT_SETTINGS = dict(default_map=ConfigFileProcessor.read_config())

    @click.command(context_settings=CONTEXT_SETTINGS)
    @click.option("-n", "--number", "numbers", type=int, multiple=True)
    @click.pass_context
    def command_with_config(ctx, numbers):
        # -- ACCESS ADDITIONAL DATA FROM CONFIG FILES: Using ctx.default_map
        for person_data_key in ctx.default_map.keys():
            if not person_data_key.startswith("person."):
                continue
            person_data = ctx.default_map[person_data_key]
            process_person_data(person_data)    # as dict.

By default the first schema listed in config_section_schemas is treated as the primary, meaning that its keys / values are merged into the top level of the default_map. This can be useful since there is then an order of precedence of variables loaded into the click command in different ways. This order is CLI > Configuration file > Environment > Default. Other schemas (not the primary), are added as a dictionary to the default_map where the key is the section name and the value is a dictionary of the keys / values in that section. These can be used manually as in the example above.

If you want to merge multiple schemas into the top level of default_map, you add them to the config_section_primary_schemas list. Understand that this will combine the namespaces, so there may be conflicts depending on usage. The advantage of this is having the precedence listed above for any listed schema. As an example:

.. code-block:: python

    class ConfigFileProcessor(ConfigFileReader):
        config_files = ["foo.ini", "foo.cfg"]
        config_section_primary_schemas = [
            ConfigSectionSchema.Foo,
            ConfigSectionSchema.Bar,
        ]
        config_section_schemas = config_section_primary_schemas + [
            ConfigSectionSchema.Baz,
        ]


CHANGES.rst
Version History
===============================================================================

Version: 0.3.0 (unreleased)
-------------------------------------------------------------------------------

DEVELOPMENT:

* Rename default branch to "main" (was: "master")

ENHANCEMENTS:

* PR #1: Expand config_searchpath for HOME directory (provided by: @rytilahti).
  HINT: Allows using: "~/.config/*.ini".

FIXED:

None.


Version: 0.2.3 (stable; 2017-09-24)
-------------------------------------------------------------------------------

LATEST STABLE RELEASE.


tasks/test.py
# -*- coding: UTF-8 -*-
"""
Invoke test tasks.
"""

from __future__ import print_function
from invoke import task, Collection
import os.path
import sys

# -- TASK-LIBRARY:
from .invoke_cleanup import cleanup_tasks, cleanup_dirs, cleanup_files


# ---------------------------------------------------------------------------
# TASKS
# ---------------------------------------------------------------------------
@task(name="all", help={
    "args": "Command line args for test run.",
})
def test_all(ctx, args="", options=""):
    """Run all tests (default)."""
    pytest_args = select_by_prefix(args, ctx.pytest.scopes)
    pytest_should_run = not args or (args and pytest_args)
    if pytest_should_run:
        pytest(ctx, pytest_args, options=options)

    behave_args = select_by_prefix(args, ctx.behave_test.scopes)
    behave_should_run = not args or (args and behave_args)
    if False and behave_should_run:
         behave(ctx, behave_args, options=options)


@task
def clean(ctx, dry_run=False):
    """Cleanup (temporary) test artifacts."""
    directories = ctx.test.clean.directories or []
    files = ctx.test.clean.files or []
    cleanup_dirs(directories, dry_run=dry_run)
    cleanup_files(files, dry_run=dry_run)


@task(name="unit")
def unittest(ctx, args="", options=""):
    """Run unit tests."""
    pytest(ctx, args, options)


@task
def pytest(ctx, args="", options=""):
    """Run unit tests."""
    args = args or ctx.pytest.args
    options = options or ctx.pytest.options
    ctx.run("pytest {options} {args}".format(options=options, args=args))


@task(help={
    "args": "Command line args for behave",
    "format": "Formatter to use (progress, pretty, ...)",
})
def behave(ctx, args="", format="", options=""):
    """Run behave tests."""
    format  = format or ctx.behave_test.format
    options = options or ctx.behave_test.options
    args = args or ctx.behave_test.args
    if os.path.exists("bin/behave"):
        behave = "{python} bin/behave".format(python=sys.executable)
    else:
        behave = "{python} -m behave".format(python=sys.executable)

    for group_args in grouped_by_prefix(args, ctx.behave_test.scopes):
        ctx.run("{behave} -f {format} {options} {args}".format(
            behave=behave, format=format, options=options, args=group_args))

@task(help={
    "args":     "Command line args for test",
    "report":   "Coverage report format to use (report, html, xml)",
})
def coverage(ctx, args="", report="report", append=False, combine=True):
    append = append or ctx.coverage.append
    combine = combine or ctx.coverage.combine
    report_formats = ctx.coverage.report_formats or []
    if report not in report_formats:
        report_formats.insert(0, report)
    opts = []
    if append:
        opts.append("--append")

    ctx.run("coverage run -m pytest {options} {args}".format(
            args=args, options=" ".join(opts)))
    if combine:
        ctx.run("coverage combine")
    for report_format in report_formats:
        ctx.run("coverage {report_format}".format(report_format=report_format))


# ---------------------------------------------------------------------------
# UTILITIES:
# ---------------------------------------------------------------------------
def select_prefix_for(arg, prefixes):
    for prefix in prefixes:
        if arg.startswith(prefix):
            return prefix
    return os.path.dirname(arg)

def select_by_prefix(args, prefixes):
    selected = []
    for arg in args.strip().split():
        assert not arg.startswith("-"), "REQUIRE: arg, not options"
        scope = select_prefix_for(arg, prefixes)
        if scope:
            selected.append(arg)
    return " ".join(selected)

def grouped_by_prefix(args, prefixes):
    """Group behave args by (directory) scope into multiple test-runs."""
    group_args = []
    current_scope = None
    for arg in args.strip().split():
        assert not arg.startswith("-"), "REQUIRE: arg, not options"
        scope = select_prefix_for(arg, prefixes)
        if scope != current_scope:
            if group_args:
                # -- DETECTED GROUP-END:
                yield " ".join(group_args)
                group_args = []
            current_scope = scope
        group_args.append(arg)
    if group_args:
        yield " ".join(group_args)


# ---------------------------------------------------------------------------
# TASK MANAGEMENT / CONFIGURATION
# ---------------------------------------------------------------------------
namespace = Collection(clean, unittest, pytest, coverage)
namespace.add_task(test_all, default=True)
namespace.configure({
    "test": {
        "clean": {
            "directories": [
                ".cache", "assets",                         # -- TEST RUNS
                "__WORKDIR__", "reports", "test_results",   # -- BEHAVE test
            ],
            "files": [
                ".coverage", ".coverage.*",
                "report.html",
                "rerun*.txt", "rerun*.featureset", "testrun*.json",
            ],
        },
    },
    "pytest": {
        "scopes":   ["tests"],
        "args":   "",
        "options": "",  # -- NOTE:  Overide in configfile "invoke.yaml"
    },
    # "behave_test": behave.namespace._configuration["behave_test"],
    "behave_test": {
        "scopes":   ["features", "issue.features"],
        "args":     "features issue.features",
        "format":   "progress",
        "options":  "",  # -- NOTE:  Overide in configfile "invoke.yaml"
    },
    "coverage": {
        "append":   False,
        "combine":  True,
        "report_formats": ["report", "html"],
    },
})

# -- ADD CLEANUP TASK:
cleanup_tasks.add_task(clean, "clean_test")
cleanup_tasks.configure(namespace.configuration())


click_configfile.py
# -*- coding: UTF-8 -*-
# (c) 2016 by Jens Engel
# pylint: disable=missing-docstring, too-few-public-methods
"""
This module provides functionality to read configuration files with commands
that already use click_ for command-line processing.

.. _click:  http://click.pocoo.org/
"""

from __future__ import absolute_import, print_function
from fnmatch import fnmatch
import os.path
import inspect
import configparser     # -- USE BACKPORT FOR: Python2
from click.types import convert_type
import six

# -----------------------------------------------------------------------------
# PACKAGE META DATA:
# -----------------------------------------------------------------------------
__version__ = "0.2.4"
__author__ = "Jens Engel"
__license__ = "BSD"


# -----------------------------------------------------------------------------
# DECORATORS
# -----------------------------------------------------------------------------
def matches_section(section_name):
    """Decorator for SectionSchema classes to define the mapping between
    a config section schema class and one or more config sections with
    matching name(s).

    .. sourcecode::

        @matches_section("foo")
        class FooSchema(SectionSchema):
            pass

        @matches_section(["bar", "baz.*"])
        class BarAndBazSchema(SectionSchema):
            pass

    .. sourcecode:: ini

        # -- FILE: *.ini
        [foo]       # USE: FooSchema
        ...

        [bar]       # USE: BarAndBazSchema
        ...

        [baz.alice] # USE: BarAndBazSchema
        ...
    """
    section_names = section_name
    if isinstance(section_name, six.string_types):
        section_names = [section_name]
    elif not isinstance(section_name, (list, tuple)):
        raise ValueError("%r (expected: string, strings)" % section_name)

    def decorator(cls):
        class_section_names = getattr(cls, "section_names", None)
        if class_section_names is None:
            cls.section_names = list(section_names)
        else:
            # -- BETTER SUPPORT: For multiple decorators
            #   @matches_section("foo")
            #   @matches_section("bar.*")
            #   class Example(SectionSchema):
            #       pass
            #   assert Example.section_names == ["foo", "bar.*"]
            approved = [name for name in section_names
                        if name not in cls.section_names]
            cls.section_names = approved + cls.section_names
        return cls
    return decorator


def assign_param_names(cls=None, param_class=None):
    """Class decorator to assign parameter name to instances of :class:`Param`.

    .. sourcecode::

        @assign_param_names
        class ConfigSectionSchema(object):
            alice = Param(type=str)
            bob   = Param(type=str)

        assert ConfigSectionSchema.alice.name == "alice"
        assert ConfigSectionSchema.bob.name == "bob"

    .. sourcecode::

        # -- NESTED ASSIGN: Covers also nested SectionSchema subclasses.
        @assign_param_names
        class ConfigSectionSchema(object):
            class Foo(SectionSchema):
                alice = Param(type=str)
                bob   = Param(type=str)

        assert ConfigSectionSchema.Foo.alice.name == "alice"
        assert ConfigSectionSchema.Foo.bob.name == "bob"
    """
    if param_class is None:
        param_class = Param

    def decorate_class(cls):
        for name, value in select_params_from_section_schema(cls, param_class,
                                                             deep=True):
            # -- ANNOTATE PARAM: By assigning its name
            if not value.name:
                value.name = name
        return cls

    # -- DECORATOR LOGIC:
    if cls is None:
        # -- CASE: @assign_param_names
        # -- CASE: @assign_param_names(...)
        return decorate_class
    else:
        # -- CASE: @assign_param_names class X: ...
        # -- CASE: assign_param_names(my_class)
        # -- CASE: my_class = assign_param_names(my_class)
        return decorate_class(cls)

# -----------------------------------------------------------------------------
# CONFIG SECTION SCHEMA DESCRIPTION
# -----------------------------------------------------------------------------
class SectionSchema(object):
    """Marker for configuration section schema classes."""
    # -- ONLY IN DERIVED CLASSES:
    # section_names = []  # Section names (or patterns) to apply SectionSchema.

    @classmethod
    def matches_section(cls, section_name, supported_section_names=None):
        """Indicates if this schema can be used for a config section
        by using the section name.

        :param section_name:    Config section name to check.
        :return: True, if this schema can be applied to the config section.
        :return: Fals, if this schema does not match the config section.
        """
        if supported_section_names is None:
            supported_section_names = getattr(cls, "section_names", None)

        # pylint: disable=invalid-name
        for supported_section_name_or_pattern in supported_section_names:
            if fnmatch(section_name, supported_section_name_or_pattern):
                return True
        # -- OTHERWISE:
        return False



class Param(object):
    """Simple Parameter class used for description config file schema params.
    Simple replacement for :class:`click.Parameter` or its variants.
    Click types are supported.

    .. sourcecode::

        class FooSchema(SectionSchema):
            name    = Param(type=str)
            flag    = Param(type=bool, default=True)
            numbers = Param(type=int, multiple=True)
            filenames = Param(type=click.Path(), multiple=True)

    .. sourcecode:: ini

        # -- FILE: foo.ini
        [foo]
        flag = yes      # -- SUPPORTS: true, false, yes, no (case-insensitive)
        name = Alice and Bob
        numbers = 1 4 9 16 25
        filenames = foo/xxx.txt
            bar/baz/zzz.txt
    """
    # pylint: disable=redefined-builtin

    def __init__(self, name=None, type=None, multiple=None, default=None,
                 help=None):
        self.name = name
        self.type = convert_type(type, default)
        self.multiple = multiple
        self.default = default
        self.help = help

    def parse(self, text):
        if self.multiple:
            parts = text.strip().split()
            values = [self.type.convert(value, self, ctx=None)
                      for value in parts]
            return values
        else:
            return self.type.convert(text, self, ctx=None)


# -----------------------------------------------------------------------------
# PARSING CONFIG SECTIONS WITH SCHEMA DESCRIPTION
# -----------------------------------------------------------------------------
def select_params_from_section_schema(section_schema, param_class=Param,
                                      deep=False):
    """Selects the parameters of a config section schema.

    :param section_schema:  Configuration file section schema to use.
    :return: Generator of params
    """
    # pylint: disable=invalid-name
    for name, value in inspect.getmembers(section_schema):
        if name.startswith("__") or value is None:
            continue    # pragma: no cover
        elif inspect.isclass(value) and deep:
            # -- CASE: class => SELF-CALL (recursively).
            # pylint: disable= bad-continuation
            cls = value
            for name, value in select_params_from_section_schema(cls,
                                            param_class=param_class, deep=True):
                yield (name, value)
        elif isinstance(value, param_class):
            yield (name, value)


def parse_config_section(config_section, section_schema):
    """Parse a config file section (INI file) by using its schema/description.

    .. sourcecode::

        import configparser     # -- NOTE: Use backport for Python2
        import click
        from click_configfile import SectionSchema, Param, parse_config_section

        class ConfigSectionSchema(object):
            class Foo(SectionSchema):
                name    = Param(type=str)
                flag    = Param(type=bool)
                numbers = Param(type=int, multiple=True)
                filenames = Param(type=click.Path(), multiple=True)

        parser = configparser.ConfigParser()
        parser.read(["foo.ini"])
        config_section = parser["foo"]
        data = parse_config_section(config_section, ConfigSectionSchema.Foo)
        # -- FAILS WITH: click.BadParameter if conversion errors occur.

    .. sourcecode:: ini

        # -- FILE: foo.ini
        [foo]
        name = Alice
        flag = yes      # true, false, yes, no (case-insensitive)
        numbers = 1 4 9 16 25
        filenames = foo/xxx.txt
            bar/baz/zzz.txt

    :param config_section:  Config section to parse
    :param section_schema:  Schema/description of config section (w/ Param).
    :return: Retrieved data, values converted to described types.
    :raises: click.BadParameter, if conversion error occurs.
    """
    storage = {}
    for name, param in select_params_from_section_schema(section_schema):
        value = config_section.get(name, None)
        if value is None:
            if param.default is None:
                continue
            value = param.default
        else:
            value = param.parse(value)
        # -- DIAGNOSTICS:
        # print("  %s = %s" % (name, repr(value)))
        storage[name] = value
    return storage


# -----------------------------------------------------------------------------
# SUPPORT: READ CONFIGFILE
# -----------------------------------------------------------------------------
def generate_configfile_names(config_files, config_searchpath=None):
    """Generates all configuration file name combinations to read.

    .. sourcecode::

        # -- ALGORITHM:
        #    First basenames/directories are prefered and override other files.
        for config_path in reversed(config_searchpath):
            for config_basename in reversed(config_files):
                config_fname = os.path.join(config_path, config_basename)
                if os.path.isfile(config_fname):
                    yield config_fname

    :param config_files:        List of config file basenames.
    :param config_searchpath:   List of directories to look for config files.
    :return: List of available configuration file names (as generator)
    """
    if config_searchpath is None:
        config_searchpath = ["."]

    for config_path in reversed(config_searchpath):
        for config_basename in reversed(config_files):
            config_fname = os.path.join(config_path, config_basename)
            config_fname = os.path.expanduser(config_fname)
            config_fname = os.path.expanduser(config_fname)
            if os.path.isfile(config_fname):
                # MAYBE: yield os.path.normpath(config_fname)
                yield config_fname


def select_config_sections(configfile_sections, desired_section_patterns):
    """Select a subset of the sections in a configuration file by using
    a list of section names of list of section name patters
    (supporting :mod:`fnmatch` wildcards).

    :param configfile_sections: List of config section names (as strings).
    :param desired_section_patterns:
    :return: List of selected section names or empty list (as generator).
    """
    for section_name in configfile_sections:
        for desired_section_pattern in desired_section_patterns:
            if fnmatch(section_name, desired_section_pattern):
                yield section_name


# -----------------------------------------------------------------------------
# BOILER-PLATE FOR CONFIG-FILE READER
# -----------------------------------------------------------------------------
class ConfigFileReader(object):
    """Generic configuration file reader.
    Concrete configuration file reader class must extend it and specify the
    concrete details.

    EXAMPLE:

    .. sourcecode::

        # -- FILE: hello_command.py
        from click_configfile import ConfigFileReader, Param, SectionSchema
        from click_configfile import matches_section

        class ConfigSectionSchema(object):

            @matches_section("hello")
            class Hello(SectionSchema):
                name = Param(type=str)

            @matches_section("hello.more.*")
            class HelloMore(SectionSchema):
                numbers = Param(type=int, multiple=True)

        class ConfigFileProcessor(ConfigFileReader):
            config_files = ["hello.ini", "hello.cfg"]
            config_section_schemas = [
                ConfigSectionSchema.Hello,     # PRIMARY SCHEMA
                ConfigSectionSchema.HelloMore,
            ]

            @classmethod
            def get_storage_name_for(cls, section_name):
                if section_name == "hello":
                    return ""   # -- MERGE-INTO-STORAGE
                elif section_name.startswith("hello.more."):
                    # -- EXAMPLE:  hello.more.alice  ->  alice
                    return section_name.replace("hello.more.", "", 1)
                # -- OTHERWISE: Delegate to BaseClass
                return ConfigFileReader.get_storage_name_for(section_name)
                # OR: raise LookupError(section_name)


        # -- COMMANDS:
        CONTEXT_SETTINGS = dict(default_map=ConfigFileProcessor.read_config())

        @click.command(context_settings=CONTEXT_SETTINGS)
        @click.option("-n", "--name", type=str)
        @click.pass_context
        def hello(ctx, name):
            pass
    """
    config_files = []               # Config filename variants (basenames).
    config_section_schemas = []     # Config section schema description.
    config_sections = []            # OPTIONAL: Config sections of interest.
    config_searchpath = ["."]       # OPTIONAL: Where to look for config files.
    config_section_primary_schemas = [] # OPTIONAL: Schemas to merge into primary.

    # -- GENERIC PART:
    # Uses declarative specification from above (config_files, config_sections, ...)
    @classmethod
    def read_config(cls):
        configfile_names = list(
            generate_configfile_names(cls.config_files, cls.config_searchpath))
        parser = configparser.ConfigParser()
        parser.optionxform = str
        parser.read(configfile_names)

        if not cls.config_sections:
            # -- AUTO-DISCOVER (once): From cls.config_section_schemas
            cls.config_sections = cls.collect_config_sections_from_schemas()

        storage = {}
        for section_name in select_config_sections(parser.sections(),
                                                   cls.config_sections):
            # print("PROCESS-SECTION: %s" % section_name)
            config_section = parser[section_name]
            cls.process_config_section(config_section, storage)
        return storage

    @classmethod
    def collect_config_sections_from_schemas(cls, config_section_schemas=None):
        # pylint: disable=invalid-name
        """Derive support config section names from config section schemas.
        If no :param:`config_section_schemas` are provided, the schemas from
        this class are used (normally defined in the DerivedClass).

        :param config_section_schemas:  List of config section schema classes.
        :return: List of config section names or name patterns (as string).
        """
        if config_section_schemas is None:
            config_section_schemas = cls.config_section_schemas

        collected = []
        for schema in config_section_schemas:
            collected.extend(schema.section_names)
            # -- MAYBE BETTER:
            # for name in schema.section_names:
            #    if name not in collected:
            #        collected.append(name)
        return collected

    # -- SPECIFIC PART:
    # Specifies which schema to use and where data should be stored.
    @classmethod
    def process_config_section(cls, config_section, storage):
        """Process the config section and store the extracted data in
        the param:`storage` (as outgoing param).
        """
        # -- CONCEPT:
        # if not storage:
        #     # -- INIT DATA: With default parts.
        #     storage.update(dict(_PERSONS={}))
        schema = cls.select_config_schema_for(config_section.name)
        if not schema:
            message = "No schema found for: section=%s"
            raise LookupError(message % config_section.name)

        # -- PARSE AND STORE CONFIG SECTION:
        section_storage = cls.select_storage_for(config_section.name, storage)
        section_data = parse_config_section(config_section, schema)
        section_storage.update(section_data)

    @classmethod
    def select_config_schema_for(cls, section_name):
        """Select the config schema that matches the config section (by name).

        :param section_name:    Config section name (as key).
        :return: Config section schmema to use (subclass of: SectionSchema).
        """
        # pylint: disable=cell-var-from-loop, redefined-outer-name
        for section_schema in cls.config_section_schemas:
            schema_matches = getattr(section_schema, "matches_section", None)
            if schema_matches is None:
                # -- OTHER SCHEMA CLASS: Reuse SectionSchema functionality.
                schema_matches = lambda name: SectionSchema.matches_section(
                    name, section_schema.section_names)

            if schema_matches(section_name):
                return section_schema
        return None

    @classmethod
    def get_storage_name_for(cls, section_name):
        """Selects where to store the retrieved data from the config section.

        :param section_name:    Config section name to process (as string).
        :return: EMPTY-STRING or None, indicates MERGE-WITH-DEFAULTS.
        :return: NON-EMPTY-STRING, for key in default_map to use.
        """
        sections_to_merge = cls.collect_config_sections_from_schemas(
            cls.config_section_primary_schemas)
        if cls.config_sections and cls.config_sections[0] == section_name:
            # -- PRIMARY-SECTION: Merge into storage (default_map).
            return ""
        elif cls.config_sections and section_name in sections_to_merge:
            # -- NON-PRIMARY-SECTION W/MERGE:  Merge into storage (default_map).
            return ""
        else:
            return section_name
        # -- MAYBE BETTER:
        # raise LookupError("Unknown section: %s" % section_name)

    @classmethod
    def select_storage_for(cls, section_name, storage):
        """Selects the data storage for a config section within the
        :param:`storage`. The primary config section is normally merged into
        the :param:`storage`.

        :param section_name:    Config section (name) to process.
        :param storage:         Data storage to use.
        :return: :param:`storage` or a part of it (as section storage).
        """
        section_storage = storage
        storage_name = cls.get_storage_name_for(section_name)
        if storage_name:
            section_storage = storage.get(storage_name, None)
            if section_storage is None:
                section_storage = storage[storage_name] = dict()
        return section_storage
